<a href="https://colab.research.google.com/github/Geethanjaliramesh/deep-learning-learning-lab/blob/main/Nipah_Inhibitor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/nipah_data_set/molecular_descriptor_training.csv")

print(df.shape)
print(df.head)

(95, 46)
<bound method NDFrame.head of             Name   AATSC5e    MATS5e      JGI9     JGI10          IC50  FP169  \
0   anti_NIV_001 -0.011084 -0.074365  0.009018  0.009241  2.880000e-08      1   
1   anti_NIV_002  0.013893  0.126755  0.009214  0.007332  3.240000e-08      1   
2   anti_NIV_003 -0.002703 -0.023649  0.010379  0.007696  5.010000e-08      1   
3   anti_NIV_004  0.005026  0.058806  0.008525  0.008948  7.080000e-08      1   
4   anti_NIV_005  0.009286  0.053779  0.006953  0.019250  1.200000e-07      1   
..           ...       ...       ...       ...       ...           ...    ...   
90  anti_NIV_091 -0.006241 -0.086565  0.007669  0.007285  6.606930e-04      0   
91  anti_NIV_092 -0.029594 -0.282971  0.000000  0.000000  6.606930e-04      1   
92  anti_NIV_093  0.000000  0.000000  0.000000  0.000000  1.023293e-03      0   
93  anti_NIV_094 -0.000392 -0.009457  0.003636  0.003306  1.862087e-03      0   
94  anti_NIV_095 -0.033646 -0.321628  0.000000  0.000000  7.585776e-03

In [5]:
print(df.columns.tolist())

['Name', 'AATSC5e', 'MATS5e', 'JGI9', 'JGI10', 'IC50', 'FP169', 'FP204', 'FP339', 'FP396', 'FP490', 'FP551', 'FP582', 'FP606', 'ExtFP79', 'ExtFP442', 'ExtFP584', 'ExtFP700', 'ExtFP1010', 'ExtFP1019', 'GraphFP158', 'GraphFP504', 'GraphFP622', 'GraphFP762', 'GraphFP860', 'GraphFP906', 'GraphFP1007', 'MACCSFP26', 'MACCSFP150', 'SubFP147', 'KRFP349', 'KRFP360', 'KRFP364', 'KRFP397', 'KRFP607', 'KRFP1538', 'KRFP2135', 'KRFP3940', 'KRFPC349', 'KRFPC2135', 'KRFPC2694', 'KRFPC3139', 'KRFPC3520', 'KRFPC4292', 'PIC50', 'Unnamed: 45']


In [6]:
y = df['PIC50']

In [7]:
x = df.drop(columns=['Name','IC50','PIC50','Unnamed: 45'])

In [8]:
print('original dataset', df.shape)
print("x shape:", x.shape)
print('y shape:',y.shape)

print("\nFeature columns:")
print(x.columns.tolist())

original dataset (95, 46)
x shape: (95, 42)
y shape: (95,)

Feature columns:
['AATSC5e', 'MATS5e', 'JGI9', 'JGI10', 'FP169', 'FP204', 'FP339', 'FP396', 'FP490', 'FP551', 'FP582', 'FP606', 'ExtFP79', 'ExtFP442', 'ExtFP584', 'ExtFP700', 'ExtFP1010', 'ExtFP1019', 'GraphFP158', 'GraphFP504', 'GraphFP622', 'GraphFP762', 'GraphFP860', 'GraphFP906', 'GraphFP1007', 'MACCSFP26', 'MACCSFP150', 'SubFP147', 'KRFP349', 'KRFP360', 'KRFP364', 'KRFP397', 'KRFP607', 'KRFP1538', 'KRFP2135', 'KRFP3940', 'KRFPC349', 'KRFPC2135', 'KRFPC2694', 'KRFPC3139', 'KRFPC3520', 'KRFPC4292']


In [9]:
from sklearn.model_selection import KFold, cross_validate
from sklearn.dummy import DummyRegressor



In [10]:
#5fold creation

cv = KFold(
    n_splits=5,
    shuffle = True,
    random_state=42
)



In [11]:
baseline = DummyRegressor(strategy="mean")

In [12]:
score = cross_validate(
    baseline,
    x,
    y,
    cv=cv,
    scoring=[
        "neg_mean_absolute_error",
        "neg_root_mean_squared_error",
        "r2"
    ]
)

In [13]:
print(score)

{'fit_time': array([0.00198746, 0.00141907, 0.00142264, 0.00167322, 0.00155902]), 'score_time': array([0.00263333, 0.002321  , 0.00212169, 0.00328755, 0.0021069 ]), 'test_neg_mean_absolute_error': array([-0.90630194, -0.94762465, -0.99132271, -0.80911357, -1.04044321]), 'test_neg_root_mean_squared_error': array([-1.08271847, -1.07159136, -1.10032223, -1.20044763, -1.23814422]), 'test_r2': array([-0.11965489, -0.00140451, -0.26632269, -0.58410276, -0.00524955])}


In [14]:
mae = -score["test_neg_mean_absolute_error"]
rmse = -score["test_neg_root_mean_squared_error"]
r2 = score["test_r2"]
print(f"Mean MAE:  {mae.mean():.3f}")
print(f"Mean RMSE: {rmse.mean():.3f}")
print(f"Mean R²:   {r2.mean():.3f}")

Mean MAE:  0.939
Mean RMSE: 1.139
Mean R²:   -0.195


Dummy Mean Baseline — 5-Fold CV

MAE  ≈ 0.939
RMSE ≈ 1.139
R²   ≈ -0.195


Our dummy baseline's predictions are wrong by about 0.94 pIC50 units on average.

Now we are going to build our first model Linear reggression
Not because we expect it to be the best model. It's because it gives us another simple benchmark


In [15]:
from sklearn.linear_model import LinearRegression


In [16]:
linear_model = LinearRegression()
linear_scores = cross_validate(
    linear_model,
    x,
    y,
    cv=cv,
    scoring=[
        "neg_mean_absolute_error",
        "neg_root_mean_squared_error",
        "r2"
    ]
)

In [18]:
linear_mae = -linear_scores["test_neg_mean_absolute_error"]
linear_rmse = -linear_scores["test_neg_root_mean_squared_error"]
linear_r2 = linear_scores["test_r2"]

print("Linear Regression Results")

print(f"Mean MAE:  {linear_mae.mean():.3f}")
print(f"Mean RMSE: {linear_rmse.mean():.3f}")
print(f"Mean R²:   {linear_r2.mean():.3f}")

print("\nMAE per fold:", linear_mae)
print("RMSE per fold:", linear_rmse)
print("R² per fold:", linear_r2)

Linear Regression Results
Mean MAE:  0.635
Mean RMSE: 0.760
Mean R²:   0.424

MAE per fold: [0.69376979 0.53195887 0.57674288 0.8439399  0.53072044]
RMSE per fold: [0.84057433 0.66785603 0.68761924 0.98718415 0.61920459]
R² per fold: [ 0.32515297  0.61102934  0.50546088 -0.07125598  0.74858004]


# Scikit learn Pipeline
now we are intoducing the pipeline also we are standardising the features


In [21]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge


In [22]:
ridge_model = Pipeline([
    ("scalar", StandardScaler()),
    ("ridge", Ridge(alpha=1.0))
])

In [23]:
ridge_scores = cross_validate(
    ridge_model,
    x,
    y,
    cv=cv,
    scoring=[
        "neg_mean_absolute_error",
        "neg_root_mean_squared_error",
        "r2"]
)

In [24]:
ridge_mae = -ridge_scores["test_neg_mean_absolute_error"]
ridge_rmse = -ridge_scores["test_neg_root_mean_squared_error"]
ridge_r2 = ridge_scores["test_r2"]

print("Ridge Regression Results")

print(f"Mean MAE:  {ridge_mae.mean():.3f}")
print(f"Mean RMSE: {ridge_rmse.mean():.3f}")
print(f"Mean R²:   {ridge_r2.mean():.3f}")

print("\nMAE per fold:", ridge_mae)
print("RMSE per fold:", ridge_rmse)
print("R² per fold:", ridge_r2)

Ridge Regression Results
Mean MAE:  0.603
Mean RMSE: 0.726
Mean R²:   0.483

MAE per fold: [0.65681738 0.5156406  0.55121479 0.76666812 0.52299665]
RMSE per fold: [0.80741274 0.64667656 0.6615043  0.90170199 0.61267972]
R² per fold: [0.37734958 0.63530872 0.54231155 0.10623578 0.7538508 ]
